File: apps/android/app/src/main/java/ai/openclaw/app/ui/chat/ChatSourcePreviews.kt
package ai.openclaw.app.ui.chat

import ai.openclaw.app.chat.ChatMessage
import ai.openclaw.app.chat.ChatRecordedSource
import ai.openclaw.app.takeUtf16Safe
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import org.commonmark.node.Code
import org.commonmark.node.Document
import org.commonmark.node.FencedCodeBlock
import org.commonmark.node.HardLineBreak
import org.commonmark.node.Image
import org.commonmark.node.IndentedCodeBlock
import org.commonmark.node.Link
import org.commonmark.node.Node
import org.commonmark.node.Paragraph
import org.commonmark.node.SoftLineBreak
import org.commonmark.node.Text

internal data class ChatSourcePreview(
  val url: String,
  val title: String,
  val domain: String,
  val excerpt: String?,
  val pageExcerpt: Boolean,
  val aliases: Set<String>,
)

internal data class ChatSourceLinkContext(
  val gatewayUrl: String? = null,
  val basePath: String = "",
  val publicOrigin: String? = null,
)

private data class SourceCandidates(
  var source: ChatRecordedSource,
  var page: ChatRecordedSource? = null,
  var search: ChatRecordedSource? = null,
)

internal fun chatSourceUrl(value: String?): HttpUrl? =
  value
    ?.takeIf { it.length <= 2_048 && (it.startsWith("https://", true) || it.startsWith("http://", true)) }
    ?.toHttpUrlOrNull()
    ?.takeIf { it.username.isEmpty() && it.password.isEmpty() }

internal fun chatSourceKey(value: String): String? =
  chatSourceUrl(value)
    ?.newBuilder()
    ?.fragment(null)
    ?.build()
    ?.toString()

/** Matches recorded, normalized web evidence to citations in one completed answer. */
internal fun extractChatSourcePreviews(
  messages: List<ChatMessage>,
  answer: ChatMessage,
  context: ChatSourceLinkContext = ChatSourceLinkContext(),
): List<ChatSourcePreview> {
  val run = answer.runId?.takeIf(String::isNotBlank) ?: return emptyList()
  if (answer.role != "assistant" || answer.phase == "commentary" || answer.isError || answer.isSyntheticDisplay || answer.provenance != null) return emptyList()
  val answerIndex = messages.indexOfFirst { it === answer }
  if (answerIndex < 0) return emptyList()
  val text =
    answer.content
      .filter { it.type == "text" }
      .mapNotNull { it.text }
      .joinToString("\n")
  if (text.isBlank() || text.length > 30_000) return emptyList()
  val links = markdownSourceLinks(text)
  if (links.isEmpty()) return emptyList()
  val cited = links.toMutableSet()
  val sources = mutableMapOf<String, SourceCandidates>()
  val redirects = mutableMapOf<String, String>()
  val calls = mutableMapOf<String, String>()
  for (message in messages.take(answerIndex)) {
    if (message.runId != run || message.role !in setOf("assistant", "toolresult") || message.isSyntheticDisplay || message.provenance != null || message.isError) continue
    if (message.sourceTools.any { it.runId != null && it.runId != run }) continue
    for (tool in message.sourceTools) {
      if (!tool.isResult) {
        if (tool.callId != null && tool.name != null) calls[tool.callId] = tool.name
        continue
      }
      if (tool.isError || !tool.completed) continue
      val name = tool.callId?.let(calls::get) ?: tool.envelopeName
      if (name !in setOf("web_search", "web_fetch") || (tool.name != null && tool.name != name) || (tool.envelopeName != null && tool.envelopeName != name)) continue
      for (source in tool.sources) {
        if (source.toolName != name) continue
        val url = chatSourceUrl(source.url) ?: continue
        val key = checkNotNull(chatSourceKey(url.toString()))
        val requestedKey = chatSourceKey(source.requestedUrl) ?: continue
        val keys = setOf(key, requestedKey)
        if (keys.none(cited::contains)) continue
        val candidates = sources.getOrPut(key) { SourceCandidates(source) }
        if (name == "web_fetch") {
          candidates.source = source
          candidates.page = source
          candidates.search = candidates.search ?: keys.firstNotNullOfOrNull { sources[it]?.search }
          cited.add(key)
          keys.filter { it != key }.forEach { redirects[it] = key }
          redirects.remove(key)
        } else if (candidates.search?.prose == null) {
          candidates.search = source
          if (candidates.page == null) candidates.source = source
        }
      }
    }
  }

  fun destination(link: String): String {
    var key = link
    val visited = mutableSetOf<String>()
    while (visited.add(key)) key = redirects[key] ?: break
    return key
  }
  val previews = linkedMapOf<String, ChatSourcePreview>()
  for (link in links) {
    if (hasDedicatedSourcePresentation(link, context)) continue
    val key = destination(link)
    val candidate = sources[key] ?: continue
    if (hasDedicatedSourcePresentation(key, context) || previews.containsKey(key)) continue
    val source = candidate.source
    val url = chatSourceUrl(source.url) ?: continue
    val pageExcerpt = candidate.page?.prose?.let { sourceExcerpt(it, page = true) }
    previews[key] =
      ChatSourcePreview(
        url = url.toString(),
        title = source.title?.let(::sourceTitle)?.takeIf(String::isNotEmpty) ?: url.host,
        domain = url.host,
        excerpt = pageExcerpt ?: candidate.search?.prose?.let { sourceExcerpt(it, page = false) },
        pageExcerpt = pageExcerpt != null,
        aliases = (redirects.keys + links + key).filter { destination(it) == key }.toSet(),
      )
    if (previews.size == 8) break
  }
  return previews.values.toList()
}

private fun markdownSourceLinks(text: String): List<String> =
  buildList {
    fun walk(start: Node?) {
      var node = start
      while (node != null) {
        when (node) {
          is Link -> node.destination?.let(::chatSourceKey)?.let(::add)
          is Code, is FencedCodeBlock, is IndentedCodeBlock, is Image -> Unit
          else -> walk(node.firstChild)
        }
        node = node.next
      }
    }
    walk(parseChatMarkdown(text).firstChild)
  }

private fun inlineSourceText(start: Node?): String =
  buildString {
    fun walk(start: Node?) {
      var node = start
      while (node != null) {
        when (node) {
          is Text -> append(node.literal)
          is Code -> append(node.literal)
          is SoftLineBreak, is HardLineBreak -> append(' ')
          is Image, is FencedCodeBlock, is IndentedCodeBlock -> Unit
          else -> walk(node.firstChild)
        }
        node = node.next
      }
    }
    walk(start)
  }.replace(Regex("\\s+"), " ").trim()

private fun sourceTitle(prose: String): String =
  buildList {
    var block = parseChatMarkdown(prose).firstChild
    while (block != null) {
      inlineSourceText(block.firstChild).takeIf(String::isNotEmpty)?.let(::add)
      block = block.next
    }
  }.joinToString(" ").takeUtf16Safe(180)

private fun sourceExcerpt(
  prose: String,
  page: Boolean,
): String? {
  fun find(start: Node?): String? {
    var node = start
    while (node != null) {
      if (node is Paragraph && (!page || node.parent is Document)) {
        val text = inlineSourceText(node.firstChild)
        if (text.length >= if (page) 60 else 1) return if (text.length > 280) text.takeUtf16Safe(279).trimEnd() + "…" else text
      }
      if (!page) find(node.firstChild)?.let { return it }
      node = node.next
    }
    return null
  }
  return find(parseChatMarkdown(prose).firstChild)
}

private fun hasDedicatedSourcePresentation(
  value: String,
  context: ChatSourceLinkContext,
): Boolean {
  val url = chatSourceUrl(value) ?: return false
  val parts = url.encodedPath.split('/').filter(String::isNotEmpty)
  if (url.scheme == "https" && url.host == "github.com" && url.port == 443 && parts.size >= 4 &&
    parts[2] in setOf("issues", "pull") && Regex("[1-9][0-9]{0,9}").matches(parts[3])
  ) {
    return true
  }
  val local =
    listOfNotNull(context.gatewayUrl, context.publicOrigin).mapNotNull(::chatSourceUrl).any {
      it.scheme == url.scheme && it.host == url.host && it.port == url.port
    }
  if (!local) return false
  val prefix = context.basePath.trimEnd('/')
  val path = url.encodedPath.trimEnd('/')
  val route = listOf("chat", "dashboard").firstOrNull { path.startsWith("$prefix/$it/") } ?: return false
  val segments = path.removePrefix("$prefix/$route/").split('/')
  return segments.all(String::isNotBlank) && (segments.getOrNull(1) != "~key" || segments.size > 2)
}

File: apps/shared/OpenClawKit/Sources/OpenClawChatUI/ChatSourcePreviews.swift
import Foundation
import Markdown
import OpenClawKit

struct ChatSourcePreview: Identifiable, Hashable {
    enum ExcerptKind: Hashable {
        case search
        case page
    }

    let url: URL
    let title: String
    let domain: String
    let excerpt: String?
    let excerptKind: ExcerptKind?
    var citedURLs: [URL] = []

    var id: String {
        self.url.absoluteString
    }

    func represents(_ url: URL) -> Bool {
        guard let canonical = ChatSourcePreviewProjector.url(url.absoluteString) else { return false }
        let key = ChatSourcePreviewProjector.key(canonical)
        return ChatSourcePreviewProjector.key(self.url) == key ||
            self.citedURLs.contains { ChatSourcePreviewProjector.key($0) == key }
    }
}

/// Keeps parsed Markdown and recorded result fields for unchanged transcript rows.
/// The projection is rebuilt from canonical history, never from a neighboring turn.
struct ChatSourcePreviewProjector {
    private struct Entry {
        let message: OpenClawChatMessage
        let links: [URL]
        let results: [Result]
    }

    private struct Result {
        let callID: String?
        let name: String?
        let candidates: [Candidate]
    }

    private struct Candidate: Hashable {
        let url: URL
        let keys: [String]
        let name: String
        let title: String?
        let excerpt: String?
    }

    private struct Sources {
        var source: Candidate
        var page: Candidate?
        var search: Candidate?

        func preview(
            cached: [Candidate: RenderedCandidate],
            retained: inout [Candidate: RenderedCandidate]) -> ChatSourcePreview
        {
            func render(_ candidate: Candidate) -> RenderedCandidate {
                let value = retained[candidate] ?? cached[candidate] ?? RenderedCandidate(
                    title: ChatSourcePreviewProjector.title(candidate.title, source: candidate.name),
                    excerpt: ChatSourcePreviewProjector.excerpt(candidate.excerpt, source: candidate.name))
                retained[candidate] = value
                return value
            }
            let source = render(self.source)
            let pageExcerpt = self.page.map(render)?.excerpt
            let excerpt = pageExcerpt ?? self.search.map(render)?.excerpt
            return ChatSourcePreview(
                url: self.source.url,
                title: source.title ?? self.source.url.host ?? self.source.url.absoluteString,
                domain: self.source.url.host ?? "",
                excerpt: excerpt,
                excerptKind: excerpt.map { _ in pageExcerpt != nil ? .page : .search })
        }
    }

    private struct RenderedCandidate {
        let title: String?
        let excerpt: String?
    }

    private var entries: [UUID: Entry] = [:]
    private var renderedCandidates: [Candidate: RenderedCandidate] = [:]

    mutating func project(
        _ messages: [OpenClawChatMessage],
        context: OpenClawChatSourceContext? = nil) -> [UUID: [ChatSourcePreview]]
    {
        var retained: [UUID: Entry] = [:]
        var retainedCandidates: [Candidate: RenderedCandidate] = [:]
        var runs: [String: [Entry]] = [:]
        var previews: [UUID: [ChatSourcePreview]] = [:]
        for message in messages {
            guard let runID = Self.nonBlank(message.transcriptRunID) else { continue }
            let entry: Entry = if let cached = self.entries[message.id], cached.message == message {
                cached
            } else {
                Entry(
                    message: message,
                    links: message.isCompletedReply ? Self.citations(in: message) : [],
                    results: Self.results(in: message, runID: runID))
            }
            retained[message.id] = entry
            if !entry.links.isEmpty {
                previews[message.id] = Self.previews(
                    links: entry.links,
                    entries: runs[runID] ?? [],
                    context: context,
                    cached: self.renderedCandidates,
                    retained: &retainedCandidates)
            }
            runs[runID, default: []].append(entry)
        }
        self.entries = retained
        self.renderedCandidates = retainedCandidates
        return previews
    }

    private static func previews(
        links: [URL],
        entries: [Entry],
        context: OpenClawChatSourceContext?,
        cached: [Candidate: RenderedCandidate],
        retained: inout [Candidate: RenderedCandidate]) -> [ChatSourcePreview]
    {
        var cited = Set(links.map(Self.key))
        var sources: [String: Sources] = [:]
        var redirects: [String: String] = [:]
        var callNames: [String: String] = [:]
        for entry in entries {
            let message = entry.message
            guard ["assistant", "tool", "toolresult", "tool_result"].contains(message.role.lowercased()),
                  !message.content.contains(where: {
                      ($0.isToolCall || $0.isToolResult) && Self.nonBlank($0.runId).map {
                          $0 != message.transcriptRunID
                      } == true
                  })
            else { continue }
            for call in message.content where call.isToolCall {
                if let id = Self.nonBlank(call.id ?? message.toolCallId), let name = Self.nonBlank(call.name) {
                    callNames[id] = name
                }
            }
            for result in entry.results {
                let envelopeName = Self.nonBlank(message.toolName)
                let invocationName = result.callID.flatMap { callNames[$0] }
                guard let name = invocationName ?? envelopeName,
                      ["web_search", "web_fetch"].contains(name),
                      (result.name ?? invocationName) == name,
                      envelopeName == nil || envelopeName == name
                else { continue }
                for candidate in result.candidates where candidate.name == name &&
                    candidate.keys.contains(where: cited.contains)
                {
                    let canonical = Self.key(candidate.url)
                    var group = sources[canonical] ?? Sources(source: candidate)
                    if name == "web_fetch" {
                        group.source = candidate
                        group.page = candidate
                        group.search = group.search ?? candidate.keys.compactMap { sources[$0]?.search }.first
                        cited.insert(canonical)
                        for key in candidate.keys where key != canonical {
                            redirects[key] = canonical
                        }
                        // A later successful fetch ends the destination's previous redirect.
                        redirects.removeValue(forKey: canonical)
                    } else if group.search?.excerpt == nil {
                        group.search = candidate
                        if group.page == nil { group.source = candidate }
                    }
                    sources[canonical] = group
                }
            }
        }
        var previews: [ChatSourcePreview] = []
        for link in links {
            guard !Self.hasDedicatedCard(link, context: context) else { continue }
            var canonical = Self.key(link)
            var visited = Set<String>()
            while let next = redirects[canonical], visited.insert(canonical).inserted {
                canonical = next
            }
            guard let source = sources[canonical], !Self.hasDedicatedCard(source.source.url, context: context)
            else { continue }
            if let index = previews.firstIndex(where: { Self.key($0.url) == canonical }) {
                previews[index].citedURLs.append(link)
            } else if previews.count < 8 {
                var preview = source.preview(cached: cached, retained: &retained)
                preview.citedURLs = [link]
                previews.append(preview)
            }
        }
        return previews
    }

    private static func results(in message: OpenClawChatMessage, runID: String) -> [Result] {
        guard message.isError != true, message.stopReason != "error", message.stopReason != "aborted" else { return [] }
        if ["tool", "toolresult", "tool_result"].contains(message.role.lowercased()) {
            let text = ChatMessageVisibleText.displayText(in: message, includeThinking: false)
            guard !ChatToolActivity.resultIsError(message.isError, text: text) else { return [] }
            return [Self.result(
                callID: message.toolCallId,
                name: message.toolName,
                details: message.details,
                text: text)]
        }
        guard message.role.lowercased() == "assistant" else { return [] }
        return message.content.filter(\.isToolResult).compactMap { block in
            guard Self.nonBlank(block.runId).map({ $0 == runID }) ?? true,
                  !ChatToolActivity.resultIsError(block.isError, text: block.text)
            else { return nil }
            return Self.result(
                callID: block.id ?? message.toolCallId,
                name: block.name ?? message.toolName,
                details: block.details,
                text: block.text ?? block.content?.stringValue)
        }
    }

    private static func result(callID: String?, name: String?, details: AnyCodable?, text: String?) -> Result {
        let payload: [String: AnyCodable]? = if let details {
            details.dictionaryValue
        } else if let text, text.utf16.count <= 100_000, let data = text.data(using: .utf8) {
            (try? JSONDecoder().decode(AnyCodable.self, from: data))?.dictionaryValue
        } else {
            nil
        }
        guard let payload,
              let external = payload["externalContent"]?.dictionaryValue,
              let source = external["source"]?.stringValue,
              external["untrusted"]?.boolValue == true, external["wrapped"]?.boolValue == true
        else { return Result(callID: Self.nonBlank(callID), name: Self.nonBlank(name), candidates: []) }
        var candidates: [Candidate] = []
        if source == "web_search" {
            let isResults = payload["kind"]?.stringValue == "results"
            let rows = isResults ? payload["results"] : payload["kind"]?
                .stringValue == "answer" ? payload["citations"] : nil
            candidates = (rows?.arrayValue ?? []).prefix(20).compactMap { value in
                guard let row = value.dictionaryValue, let url = Self.url(row["url"]?.stringValue) else { return nil }
                return Candidate(
                    url: url,
                    keys: [Self.key(url)],
                    name: source,
                    title: row["title"]?.stringValue,
                    excerpt: isResults ? row["snippet"]?.stringValue : nil)
            }
        } else if source == "web_fetch", let status = payload["status"]?.doubleValue,
                  status >= 200, status < 300,
                  let requested = Self.url(payload["url"]?.stringValue),
                  let final = Self.url(payload["finalUrl"]?.stringValue)
        {
            candidates = [Candidate(
                url: final,
                keys: [Self.key(requested), Self.key(final)],
                name: source,
                title: payload["title"]?.stringValue,
                excerpt: payload["text"]?.stringValue)]
        }
        return Result(callID: Self.nonBlank(callID), name: Self.nonBlank(name), candidates: candidates)
    }

    private static func citations(in message: OpenClawChatMessage) -> [URL] {
        let text = ChatMessageVisibleText.visibleText(in: message)
        guard !text.isEmpty, text.utf16.count <= 30000 else { return [] }
        return chatPreviewURLs(in: text).compactMap { Self.url($0.absoluteString) }
    }

    private static func prose(_ value: String?, source: String) -> String? {
        guard let value, value.utf16.count <= 100_000 else { return nil }
        // Only the provenance-owned frame is removed; quoted markers remain source prose.
        let pattern = #"(?:^|\n)<<<EXTERNAL_UNTRUSTED_CONTENT id="([a-f0-9]{16})">>>\r?\n"# +
            #"Source: ([^\r\n]+)\r?\n---\r?\n([\s\S]*?)\r?\n"# +
            #"<<<END_EXTERNAL_UNTRUSTED_CONTENT id="\1">>>"#
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: value, range: NSRange(value.startIndex..., in: value)),
              let nameRange = Range(match.range(at: 2), in: value),
              value[nameRange] == (source == "web_search" ? "Web Search" : "Web Fetch"),
              let bodyRange = Range(match.range(at: 3), in: value)
        else { return nil }
        return String(value[bodyRange].trimmingCharacters(in: .whitespacesAndNewlines).prefix(30000))
    }

    private static func inlineText(_ markup: any Markup) -> String {
        if let text = markup as? Markdown.Text { return text.string }
        if let code = markup as? InlineCode { return code.code }
        if markup is SoftBreak || markup is LineBreak { return " " }
        if markup is Markdown.Image || markup is CodeBlock || markup is HTMLBlock || markup is InlineHTML { return "" }
        return markup.children.map(Self.inlineText).joined()
    }

    private static func cleanText(_ text: String) -> String {
        text.split(whereSeparator: \.isWhitespace).joined(separator: " ")
    }

    private static func title(_ value: String?, source: String) -> String? {
        guard let prose = Self.prose(value, source: source) else { return nil }
        let text = Self.cleanText(Document(parsing: prose).children.map(Self.inlineText).joined(separator: " "))
        return text.isEmpty ? nil : String(text.prefix(180))
    }

    private static func excerpt(_ value: String?, source: String) -> String? {
        guard let prose = Self.prose(value, source: source) else { return nil }
        let document = Document(parsing: prose)
        func paragraphs(_ markup: any Markup) -> [String] {
            if markup is Paragraph { return [Self.cleanText(Self.inlineText(markup))] }
            return markup.children.flatMap(paragraphs)
        }
        let paragraphs = source == "web_fetch"
            ? document.children.compactMap { $0 is Paragraph ? Self.cleanText(Self.inlineText($0)) : nil }
            : paragraphs(document)
        guard let text = paragraphs.first(where: { $0.count >= (source == "web_fetch" ? 60 : 1) }) else { return nil }
        return text.count > 280 ? String(text.prefix(279)).trimmingCharacters(in: .whitespacesAndNewlines) + "…" : text
    }

    private static func nonBlank(_ value: String?) -> String? {
        guard let value = value?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else { return nil }
        return value
    }

    fileprivate static func url(_ value: String?) -> URL? {
        guard let value, value.utf16.count <= 2048,
              var components = URLComponents(string: value),
              let scheme = components.scheme?.lowercased(), ["http", "https"].contains(scheme),
              let host = components.host, !host.isEmpty,
              components.user == nil, components.password == nil
        else { return nil }
        components.scheme = scheme
        components.host = host.lowercased()
        if components.path.isEmpty { components.path = "/" }
        if components.port == (scheme == "https" ? 443 : 80) { components.port = nil }
        return components.url
    }

    fileprivate static func key(_ url: URL) -> String {
        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)!
        components.fragment = nil
        return components.string!
    }

    private static func hasDedicatedCard(_ url: URL, context: OpenClawChatSourceContext?) -> Bool {
        if self.isGitHubItem(url) { return true }
        guard let context, [context.gatewayURL, context.publicOrigin].compactMap(\.self).contains(where: {
            $0.scheme?.lowercased() == url.scheme?.lowercased() && $0.host?.lowercased() == url.host?.lowercased() &&
                ($0.port ?? ($0.scheme == "https" ? 443 : 80)) == (url.port ?? (url.scheme == "https" ? 443 : 80))
        }) else { return false }
        let base = context.basePath.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let prefix = base.isEmpty ? "" : "/" + base
        let path = URLComponents(url: url, resolvingAgainstBaseURL: false)?.percentEncodedPath ?? ""
        for namespace in ["chat", "dashboard"] {
            let route = prefix + "/" + namespace + "/"
            guard path.hasPrefix(route) else { continue }
            var suffix = String(path.dropFirst(route.count))
            if suffix.hasSuffix("/") { suffix.removeLast() }
            let segments = suffix.components(separatedBy: "/")
            guard segments.allSatisfy({ !($0.removingPercentEncoding ?? "").isEmpty }) else { return false }
            return segments.count < 2 || segments[1] != "~key" || segments.count > 2
        }
        return false
    }

    private static func isGitHubItem(_ url: URL) -> Bool {
        guard url.scheme == "https", url.host == "github.com", url.port == nil else { return false }
        let parts = url.path.split(separator: "/")
        return parts.count >= 4 && ["issues", "pull"].contains(parts[2]) &&
            parts[3].range(of: #"^[1-9][0-9]{0,9}$"#, options: .regularExpression) != nil
    }
}
